number_theory/numeric.h

```diff
@@ -43,6 +45,54 @@ std::pair<T, T> exgcd(T a, T b) {
   return std::make_pair(xa, ya);
 }
 
+// Accumulates values according to the binary representation of |binary|.
+// For each bit of |binary|, from lower to higher, it calls |update| to
+// accumulate the values. The initial value is |init|.
+template <class T, class U>
+U binary_accumulate(T binary,
+                    const U &init,
+                    std::function<void(bool, U &)> update) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && !std::numeric_limits<U>::is_signed,
+      "binary_accumulate argument |n| must be an unsigned integer");
+
+  T current = binary;
+  U result = init;
+  while (current != 0) {
+    bool bit = (current % 2) != 0;
+    current /= 2;
+    update(bit, result);
+  }
+  return result;
+}
+
+// Computes the value of |base| raised to an integer power |exponent|.
+// The time complexity is O(log |exponent|).
+template <class T, class U>
+T ipow(const T &base, U exponent) {
+  static_assert(std::numeric_limits<U>::is_integer,
+                "ipow exponent must be an integer");
+
+  typedef std::pair<T, T> T2;
```

Comment:
Let's prefer the modern `using` statements over typedefs such as:
```cpp
using unsigned_U = std::make_unsigned_t<U>;
```
However, starting with lowercase might cause some confusion since it looks like a variable name. Maybe we should name it as `Unsigned_U`?

I also think `T2` is too confusing and not worth it, I would rather use `std::pair<T, T>` explicitly. If you prefer using a `using` statement for that one, maybe we can try names like `T_Pair`, `Pair_TT`, `T_PAIR`. However, the last one looks like a macro name and I think we should avoid that one. I still think going explicitly `std::pair<T, T>` is the best option.

Reply:
done